File: javaScript/longestSubstringWithoutRepeatingCharacters.js
// 3. Longest Substring Without Repeating Characters  QuestionEditorial Solution  My Submissions
// Total Accepted: 178499
// Total Submissions: 775857
// Difficulty: Medium
// Given a string, find the length of the longest substring without repeating characters.

// Examples:

// Given "abcabcbb", the answer is "abc", which the length is 3.

// Given "bbbbb", the answer is "b", with the length of 1.

// Given "pwwkew", the answer is "wke", with the length of 3. Note that the answer must be a substring, "pwke" is a subsequence and not a substring.

// Subscribe to see which companies asked this question

// Hide Tags Hash Table Two Pointers String
// Hide Similar Problems (H) Longest Substring with At Most Two Distinct Characters
// Have you met this question in a real interview? Yes  No
// Discuss Pick One

/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
	//corner case
	if (s === null || s.length < 1) {
		return 0
	}

	var left = 0;
	var n = s.length;
	var map = new Map();
	var maxL = 0;

	while (left < n) {
		if (map.size > maxL) {
			maxL = map.size;
		}
		if (!map.has(s.charAt(left))) {
			map.set(s.charAt(left), left);
			left++;
		} else {
			//map.set(s.charAt(left), true);
			left = map.get(s.charAt(left)) + 1;
			map.clear();
		}
	}
	if (map.size > maxL) {
		maxL = map.size;
	}
	return maxL; 

    
};

// 读完题后的第一想法，就是从左开始维护做指针和hashmap，如果hashmap里面有了，就把左指针更新到当前
// 做完这题简直被打击到了……逻辑各种问题 T_T, 代码也很丑

File: python/LongestSubstringWithoutRepeatingCharacters.py
# Longest Substring Without Repeating Characters

# Description
# Given a string, find the length of the longest substring without repeating characters.

# Example
# For example, the longest substring without repeating letters for "abcabcbb" is "abc", which the length is 3.

# For "bbbbb" the longest substring is "b", with the length of 1.

# Challenge 
# O(n) time

# Tags 
# String Two Pointers Hash Table
# Related Problems 
# Medium Longest Substring with At Most K Distinct Characters 22 %
class Solution:
    # @param s: a string
    # @return: an integer
    def lengthOfLongestSubstring(self, s):
        # write your code here
        if s == None or len(s) < 1:
            return 0
        map = {}
        maxL = 1
        i = 0
        
        while i < len(s):
            if len(map) >= maxL:
                maxL = len(map)
            if map.has_key(s[i]) == False:
                char = s[i]
                map[char] = True
                i += 1
            else:
                map={}  #本题的逻辑错误在这里，一旦发现重复，不应该把hash清零，而应该去掉第一个加入hash里面的值，再接着看，所以本题不能用hash单用hash做，还得加一个指针
                i += 1

        return max(maxL, len(map))

# 上述解法对aaaab这个case的计算会出错，只会计算出1， 而算不出2，逻辑错误见注释出

class Solution:
    # @param s: a string
    # @return: an integer
    def lengthOfLongestSubstring(self, s):
        # write your code here
        if s == None or len(s) < 1:
            return 0
        map = {}
        maxL = 1
        i = 0
        j = 0
        
        while i < len(s):
            if len(map) >= maxL:
                maxL = len(map)
            if map.has_key(s[i]) == False:
                char = s[i]
                map[char] = True
                i += 1
            else:
            	tmp = s[j]
                del map[tmp]
                j += 1

        return max(maxL, len(map))
# 以上解法已AC